File: app/src/main/java/net/yupol/transmissionremote/app/preferences/ServersFragment.java
package net.yupol.transmissionremote.app.preferences;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ListView;
import android.widget.RadioButton;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.fragment.app.ListFragment;

import net.yupol.transmissionremote.app.R;
import net.yupol.transmissionremote.app.TransmissionRemote;
import net.yupol.transmissionremote.app.server.Server;

import java.util.Locale;

public class ServersFragment extends ListFragment {

    private TransmissionRemote app;

    private OnServerSelectedListener serverSelectedListener;

    public ServersFragment() {
        setListAdapter(new BaseAdapter() {
            @Override
            public int getCount() {
                return app.getServers().size();
            }

            @Override
            public Server getItem(int position) {
                return app.getServers().get(position);
            }

            @Override
            public long getItemId(int position) {
                return position;
            }

            @Override
            public View getView(int position, View convertView, ViewGroup parent) {
                View itemView = convertView;
                if (itemView == null) {
                    LayoutInflater li = LayoutInflater.from(parent.getContext());
                    itemView = li.inflate(R.layout.server_item_layout, parent, false);
                }

                Server server = getItem(position);

                TextView nameText = itemView.findViewById(R.id.name);
                nameText.setText(server.getName());

                TextView addressText = itemView.findViewById(R.id.address);
                int port = server.getPort();
                String url = String.format(Locale.ROOT, "%s://%s%s",
                        server.useHttps() ? "https" : "http",
                        server.getHost(),
                        port >= 0 ? (":" + port) : "");
                addressText.setText(url);

                RadioButton radioButton = itemView.findViewById(R.id.radio_button);
                radioButton.setChecked(server.equals(app.getActiveServer()));

                return itemView;
            }
        });
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setEmptyText(getString(R.string.servers_empty_text));
    }

    @Override
    public void onResume() {
        super.onResume();
        ((BaseAdapter) getListAdapter()).notifyDataSetChanged();
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        app = (TransmissionRemote) requireActivity().getApplication();
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.servers_menu, menu);
    }

    @Override
    public void onListItemClick(@NonNull ListView l, @NonNull View v, int position, long id) {
        BaseAdapter adapter = (BaseAdapter) getListAdapter();
        Server server = (Server) adapter.getItem(position);
        if (!server.equals(app.getActiveServer())) {
            app.setActiveServer(server);
        }
        ((BaseAdapter) getListAdapter()).notifyDataSetChanged();

        if (serverSelectedListener != null) {
            serverSelectedListener.onServerSelected(server);
        }
    }

    public void setOnServerSelectedListener(OnServerSelectedListener listener) {
        serverSelectedListener = listener;
    }

    public interface OnServerSelectedListener {
        void onServerSelected(Server server);
    }
}

File: app/src/main/java/net/yupol/transmissionremote/app/preferences/ServersFragment.kt
package net.yupol.transmissionremote.app.preferences

import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.*
import android.view.View.GONE
import android.view.View.VISIBLE
import android.widget.Toast
import butterknife.BindView
import butterknife.ButterKnife
import com.mikepenz.google_material_typeface_library.GoogleMaterial
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import io.reactivex.functions.BiFunction
import net.yupol.transmissionremote.app.BaseFragment
import net.yupol.transmissionremote.app.R
import net.yupol.transmissionremote.app.TransmissionRemote
import net.yupol.transmissionremote.app.utils.DividerItemDecoration
import net.yupol.transmissionremote.app.utils.IconUtils
import net.yupol.transmissionremote.domain.model.Server
import net.yupol.transmissionremote.domain.repository.ServerRepository
import javax.inject.Inject

class ServersFragment : BaseFragment(), ServerListAdapter.OnServerSelectedListener {

    @BindView(R.id.recycler_view) lateinit var recyclerView: RecyclerView
    @BindView(R.id.empty_view) lateinit var emptyView: View

    @Inject lateinit var repo: ServerRepository

    private lateinit var adapter: ServerListAdapter
    private var serversSubscription: Disposable? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        TransmissionRemote.getInstance().appComponent().inject(this)
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)

        adapter = ServerListAdapter(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.server_list_layout, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        ButterKnife.bind(this, view)

        recyclerView.layoutManager = LinearLayoutManager(context)
        recyclerView.addItemDecoration(DividerItemDecoration(context))
        recyclerView.adapter = adapter

        subscribeToServers()
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater!!.inflate(R.menu.servers_menu, menu)
        IconUtils.setMenuIcon(activity, menu, R.id.action_add, GoogleMaterial.Icon.gmd_add)
    }

    override fun onDestroy() {
        serversSubscription?.dispose()
        super.onDestroy()
    }

    override fun onServerSelected(server: Server) {
        repo.setActiveServer(server)
    }

    private fun subscribeToServers() {
        serversSubscription?.dispose()

        serversSubscription = Observable.combineLatest(repo.servers(), repo.activeServer(), BiFunction { servers: List<Server>, activeServer: Server -> servers to activeServer })
                .subscribe { (servers, activeServer) ->
                    adapter.setServers(servers, activeServer)

                    Toast.makeText(context, activeServer.name, Toast.LENGTH_SHORT).show()

                    emptyView.visibility = if (servers.isEmpty()) VISIBLE else GONE
                }
    }
}
